services/web/server/src/simcore_service_webserver/projects/_nodes_api.py

```diff
@@ -86,24 +108,113 @@ def guess_mimetype_if_undefined(cls, values):
-            # NOTE: mimetypes.guess_type works differently in our image, therefore made it nullable if
-            # cannot guess type
-            # SEE https://github.com/ITISFoundation/osparc-simcore/issues/4385
-            values["mimetype"] = _type
+            values["mimetype"] = _guess_mimetype_from_name(file_url)
 
         return values
 
 
-async def fake_screenshots_factory(
-    request: web.Request, node_id: NodeID, node: Node
+def __get_search_key(meta_data: FileMetaDataGet) -> str:
+    return f"{meta_data.file_id}".lower()
+
+
+class _FileWithThumbnail(NamedTuple):
+    file: FileMetaDataGet
+    thumbnail: FileMetaDataGet
+
+
+def _get_files_with_thumbnails(
+    assets_files: list[FileMetaDataGet],
+) -> list[_FileWithThumbnail]:
+    """returns a list of tuples where the second entry is the thumbnails"""
+
+    selected_file_entries: dict[str, FileMetaDataGet] = {
+        __get_search_key(f): f
+        for f in assets_files
+        if not f.file_id.endswith(".hidden_do_not_remove")
+        and Path(f.file_id).suffix.lower() in _SUPPORTED_PREVIEW_FILE_EXTENSIONS
+    }
+
+    with_thumbnail_image: list[_FileWithThumbnail] = []
+
+    for selected_file in set(selected_file_entries.keys()):
+        # search for thumbnail
+        thumbnail: FileMetaDataGet | None = None
+        for extension in _SUPPORTED_THUMBNAIL_EXTENSIONS:
+            thumbnail_search_file_name = f"{selected_file}{extension}"
+            if thumbnail_search_file_name in selected_file_entries:
+                thumbnail = selected_file_entries[thumbnail_search_file_name]
+                break
+        if not thumbnail:
+            continue
+
+        # since there is a thumbnail it can be associated to a file
+        with_thumbnail_image.append(
+            _FileWithThumbnail(
+                file=selected_file_entries[selected_file],
+                thumbnail=thumbnail,
+            )
+        )
+        # remove entries which have been used
+        selected_file_entries.pop(
+            __get_search_key(selected_file_entries[selected_file]), None
+        )
+        selected_file_entries.pop(__get_search_key(thumbnail), None)
+
+    no_thumbnail_image: list[_FileWithThumbnail] = [
+        _FileWithThumbnail(file=x, thumbnail=x) for x in selected_file_entries.values()
+    ]
+
+    return with_thumbnail_image + no_thumbnail_image
+
+
+async def __get_link(
+    app: web.Application, user_id: UserID, file_meta_data: FileMetaDataGet
+) -> tuple[str, HttpUrl]:
+    return __get_search_key(file_meta_data), await get_download_link(
+        app,
+        user_id,
+        parse_obj_as(SimCoreFileLink, {"store": "0", "path": file_meta_data.file_id}),
+    )
+
+
+async def _get_node_screenshots_parallel(
+    app: web.Application,
+    user_id: UserID,
+    files_with_thumbnails: list[_FileWithThumbnail],
 ) -> list[NodeScreenshot]:
-    """
-    ONLY for testing purposes
+    """resolves links in parallel before returning all NodeScreenshot"""
```

Comment:
it's not parallel but concurrent. and as the client user of that function I do not think I should care? there is no other option anyway right?